File: tests/kotlin/tests-kotlin-gradle-junit5-allure3-playwright/src/test/java/pages/RegisterPage.java
package pages;

import com.microsoft.playwright.Locator;
import com.microsoft.playwright.Page;
import io.qameta.allure.Step;

public class RegisterPage {

    private final Page page;
    public final Locator registerForm;
    public final Locator loginInput;
    public final Locator passwordInput;
    public final Locator confirmPasswordInput;
    public final Locator submitButton;
    public final Locator formTitle;
    public final Locator errorMessage;
    public final Locator loginLink;

    public RegisterPage(Page page) {
        this.page = page;
        this.registerForm = page.getByTestId("register-form");
        this.loginInput = page.getByTestId("register-login-input");
        this.passwordInput = page.getByTestId("register-password-input");
        this.confirmPasswordInput = page.getByTestId("confirm-password-input");
        this.submitButton = page.getByTestId("register-submit-button");
        this.formTitle = page.getByTestId("register-form-title");
        this.errorMessage = page.getByTestId("register-error-message");
        this.loginLink = page.getByTestId("login-link");
    }

    @Step("Open register page")
    public RegisterPage open() {
        page.navigate("register");
        return shouldBeOpen();
    }

    @Step("Verify register page is open")
    public RegisterPage shouldBeOpen() {
        registerForm.waitFor();
        return this;
    }

    @Step("Verify register form is mounted")
    public RegisterPage shouldShowRegisterForm() {
        formTitle.waitFor();
        loginInput.waitFor();
        passwordInput.waitFor();
        confirmPasswordInput.waitFor();
        submitButton.waitFor();
        return this;
    }

    @Step("Sign up as {username}")
    public RegisterPage signup(String username, String password) {
        return signup(username, password, password);
    }

    @Step("Sign up as {username}")
    public RegisterPage signup(String username, String password, String confirmPassword) {
        loginInput.fill(username);
        passwordInput.fill(password);
        confirmPasswordInput.fill(confirmPassword);
        submitButton.click();
        return this;
    }

    @Step("Type username: {username}")
    public RegisterPage typeUsername(String username) {
        loginInput.fill(username);
        return this;
    }

    @Step("Type password")
    public RegisterPage typePassword(String password) {
        passwordInput.fill(password);
        return this;
    }

    @Step("Type confirm password")
    public RegisterPage typeConfirmPassword(String password) {
        confirmPasswordInput.fill(password);
        return this;
    }

    @Step("Submit register form expecting validation error")
    public RegisterPage submitExpectingError() {
        submitButton.click();
        errorMessage.waitFor();
        return this;
    }

    @Step("Click Login link under the register form")
    public RegisterPage clickLoginLink() {
        loginLink.click();
        return this;
    }
}

File: tests/python/tests-python-playwright/pages/register_page.py
from __future__ import annotations

import allure
from playwright.sync_api import Locator, Page


class RegisterPage:
    def __init__(self, page: Page) -> None:
        self.page = page
        self.register_form: Locator = page.get_by_test_id("register-form")
        self.login_input: Locator = page.get_by_test_id("register-login-input")
        self.password_input: Locator = page.get_by_test_id("register-password-input")
        self.confirm_password_input: Locator = page.get_by_test_id("confirm-password-input")
        self.submit_button: Locator = page.get_by_test_id("register-submit-button")
        self.form_title: Locator = page.get_by_test_id("register-form-title")
        self.error_message: Locator = page.get_by_test_id("register-error-message")
        self.login_link: Locator = page.get_by_test_id("login-link")

    @allure.step("Open register page")
    def open(self) -> RegisterPage:
        self.page.goto("register")
        return self.should_be_open()

    @allure.step("Verify register page is open")
    def should_be_open(self) -> RegisterPage:
        self.register_form.wait_for()
        return self

    @allure.step("Verify register form is mounted")
    def should_show_register_form(self) -> RegisterPage:
        self.form_title.wait_for()
        self.login_input.wait_for()
        self.password_input.wait_for()
        self.confirm_password_input.wait_for()
        self.submit_button.wait_for()
        return self

    @allure.step("Sign up as {username}")
    def signup(self, username: str, password: str, confirm_password: str | None = None) -> RegisterPage:
        confirm = password if confirm_password is None else confirm_password
        self.login_input.fill(username)
        self.password_input.fill(password)
        self.confirm_password_input.fill(confirm)
        self.submit_button.click()
        return self

    @allure.step("Type username: {username}")
    def type_username(self, username: str) -> RegisterPage:
        self.login_input.fill(username)
        return self

    @allure.step("Type password")
    def type_password(self, password: str) -> RegisterPage:
        self.password_input.fill(password)
        return self

    @allure.step("Type confirm password")
    def type_confirm_password(self, password: str) -> RegisterPage:
        self.confirm_password_input.fill(password)
        return self

    @allure.step("Submit register form expecting validation error")
    def submit_expecting_error(self) -> RegisterPage:
        self.submit_button.click()
        self.error_message.wait_for()
        return self

    @allure.step("Click Login link under the register form")
    def click_login_link(self) -> RegisterPage:
        self.login_link.click()
        return self
